linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala

```diff
@@ -61,28 +58,121 @@ object ZipkinTrace {
   val ZipkinSampleHeader = "x-b3-sampled"
   val ZipkinFlagsHeader = "x-b3-flags"
 
+  /**
+   * The separate "b3" header in b3 single header format :
+   * b3={x-b3-traceid}-{x-b3-spanid}-{if x-b3-flags 'd' else x-b3-sampled}-{x-b3-parentspanid},
+   * where the last two fields are optional.
+   */
+  val ZipkinB3SingleHeader = "b3"
+
+  /**
+   * Handle b3 the same way as finagle handles b3, extract the traceid from "b3" header, write back "X-B3-*" multi
+   * headers, and let the flow continue as if it has received "x-b3-".
+   * @note Copied/adapted from finagle TraceInfo.scala
+   */
+  def convertB3SingleHeaderToMultiHeaders(headers: HeaderMap): Unit = {
+    if (headers.contains(ZipkinB3SingleHeader)) {
+      def handleSampled(headers: HeaderMap, value: String): Unit = {
+        value match {
+          case "0" =>
+            headers.set(ZipkinSampleHeader, "0")
+          case "d" =>
+            headers.set(ZipkinFlagsHeader, "1")
+          case "1" =>
+            headers.set(ZipkinSampleHeader, "1")
+          case _ =>
+            () // do not set anything on invalid value
+        }
+        ()
+      }
+
+      def handleTraceAndSpanIds(headers: HeaderMap, a: String, b: String): Unit = {
+        headers.set(ZipkinTraceHeader, a)
+        headers.set(ZipkinSpanHeader, b)
+        ()
+      }
+
+      caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-")) match {
+        case Some(a) =>
+          a.size match {
```

Comment:
TIOLI: I know that I said that toList was not needed and it is not if you are simply matching on the size. But dont you think it would be a bit more idiomatic to do cons matches? Then you can have the extracted parts of the list be named a lot more explicitly instead of needing to index into the array?  And generally matching on size when we have all the bells and whistles to do it in more expressive ways is a bit of a code smell.

Reply:
redone it using list, see my other PR.